src/ngResource/resource.js

```diff
@@ -43,6 +52,115 @@ function shallowClearAndCopy(src, dst) {
   return dst;
 }
 
+function parseTemplateParams(template) {
+  var urlParams = Object.create(null);
+  forEach(template.split(/\W/), function(segment) {
+    if (segment === 'hasOwnProperty') {
+      throw $resourceMinErr('badname', 'hasOwnProperty is not a valid parameter name.');
+    }
+    // ignore segments that look like port numbers (e.g. `:8080`)
+    if (segment && !ONLY_DIGITS_REGEX.test(segment) && containsParam(template, segment)) {
+      // does this param appear in the query part of the URL?
+      var isQueryParamValue = (new RegExp('\\?.*=:' + segment + '(?:\\W|$)')).test(template);
+      urlParams[segment] = { isQueryParamValue: isQueryParamValue };
+    }
+  });
+  return urlParams;
+}
+
+var $composeResourceUrlFactory = ['$httpParamSerializer', function($httpParamSerializer) {
```

Comment:
Hmm, this is a real pain. My first attempt left the delegation in place but it was really ugly, since the `composeResourceUrl()` function had to return an object which contained the partially composed URL plus an object of the additional parameters. This was made even worse by the fact that if the URL template contained a query string, you would end up with some params in the URL and some in the extra params hash.

I'll change to use the `$http.defaults.paramSerializer` and add a BC notice.

Reply:
One possible alternative is to have an internal version of `composeResourceUrl` which is used by both `$resource` and `$composeResourceUrl`. This returns the params hash so that `$resource` can continue to work identically to before. Then the `$composeResourceUrl` could allow the developer to specify the param serializer in the `options` parameter.

What do you think?

Reply:
I am fine with both options (except that if we go with the first one, we should also respect an action or resource specific paramSerializer).

Reply:
If there's an internal version of composeResourceUrl, wouldn't that make it impossible to replace composeResourceUrl service so that the changes to it are picked up by $resource? Wasn't this one of the goals in this PR?

Reply:
Reading back through the original issue, that was never one of the requested use cases. It just was one of the side-benefits of having a service. I am happier to remove this benefit while maintaining backward compatibility and exposing a useful service to people who want to use it.